src/tests/utils/preprocess_asm.rs

```diff
@@ -35,29 +38,108 @@ pub const PRINT_PREFIX: &str = "L:";
 pub const PRINT_REG_PREFIX: &str = "R:";
 pub const PRINT_PTR_PREFIX: &str = "P:";
 
+pub type TemplateDictionary<'a> = HashMap<&'a str, &'a str>;
+
 /// Replaces special directives in asm with TestingTracer compatible "commands"
-pub fn preprocess_asm(asm: &str) -> String {
-    let mut result = asm.to_owned();
-    for directive in [
+pub fn preprocess_asm(
+    asm: String,
+    additional_contracts: Option<&Vec<(H160, Vec<[u8; 32]>)>>,
+    dictionary: Option<&TemplateDictionary>,
+) -> String {
+    let asm = replace_tags_in_template(asm, dictionary);
+
+    let result = [
         Directive::Print(PrintType::Text),
         Directive::Print(PrintType::Register),
         Directive::Print(PrintType::Pointer),
         Directive::Revert,
-    ] {
-        result = preprocess_directive(&result, directive);
+    ]
+    .iter()
+    .fold(asm, |acc, x| preprocess_directive(acc, *x));
+
+    link_additional_contracts(&result, additional_contracts)
+}
+
+/// Replaces tags like "${TAG}" with values from dictionary
+fn replace_tags_in_template(
+    asm_template: String,
+    dictionary: Option<&TemplateDictionary>,
+) -> String {
+    let mut result = asm_template.clone();
+    let template_regex = Regex::new(r#"\$\{[^\}]+\}"#).expect("Invalid regex");
+
+    for (_, matched) in asm_template.match_indices(&template_regex) {
+        let prefix = "${";
+        let suffix = "}";
+        let key_to_replace = matched
+            .strip_prefix(prefix)
+            .expect("Invalid text in template")
+            .strip_suffix(suffix)
+            .expect("Invalid text in template");
+
+        if dictionary == None {
+            panic!("Unknown key: {key_to_replace}");
```

Comment:
nit: (can be done in separate PR) - we seem to have too many panics. Instead we should start returning Result<String, String> - so that the caller can decide on whether they want to panic.